src/HIE/Bios/Config.hs

```diff
@@ -43,11 +65,45 @@ data CradleConfig a =
 data Callable = Program FilePath | Command String
     deriving (Show, Eq)
 
+data CabalType
+    = CabalType_ { _cabalComponent :: !(MLast String) }
+    deriving (Eq)
+
+instance Semigroup CabalType where
+    CabalType_ cr <> CabalType_ cl = CabalType_ (cr <> cl)
+
+instance Monoid CabalType where
+    mempty = CabalType_ mempty
+
+pattern CabalType :: Maybe String -> CabalType
+pattern CabalType { cabalComponent } = CabalType_ (MLast cabalComponent)
+{-# COMPLETE CabalType #-}
+
+instance Show CabalType where
+  show = show . Cabal
+
+data StackType
+    = StackType_ { _stackComponent :: !(MLast String) , _stackYaml :: !(MLast String) }
+    deriving (Eq)
+
+instance Semigroup StackType where
+    StackType_ cr yr <> StackType_ cl yl = StackType_ (cr <> cl) (yr <> yl)
+
+instance Monoid StackType where
+    mempty = StackType_ mempty mempty
+
+pattern StackType :: Maybe String -> Maybe String -> StackType
+pattern StackType { stackComponent, stackYaml } = StackType_ (MLast stackComponent) (MLast stackYaml)
+{-# COMPLETE StackType #-}
+
+instance Show StackType where
+  show = show . Stack
+
 data CradleType a
-    = Cabal { component :: Maybe String }
-    | CabalMulti [ (FilePath, String) ]
-    | Stack { component :: Maybe String }
-    | StackMulti [ (FilePath, String) ]
+    = Cabal { cabalType :: !CabalType }
+    | CabalMulti { defaultCabal :: !CabalType, subCabalComponents :: [ (FilePath, CabalType) ] }
```

Comment:
This also allows the config file:
``` yaml
cradle:
  cabal:
    components:
      - path: "./src"
        component: "lib:hie-bios"
      - path: "./"
        component: "lib:hie-bios"
```

While I dont expect any harm of it (and maybe we want soonish a similar field for cabal) but I just wanted to document it.